Feat/cuzk/reproduce on cpu

moven0831 opened the pull request:
This PR introduces the initial draft of the **GPU pipeline reproduced purely on CPU**.

Currently, the implementation accurately reproduces Multi-Scalar Multiplication (MSM) results—validated against Arkworks' MSM—when the scalar values satisfy the condition:

```rust
ScalarField::from(BigUint::from_str(SCALAR).unwrap())
```

Specifically, correctness is guaranteed only for `SCALAR` values less than or equal to 7 (`SCALAR <= 7`). Further work will focus on extending this limitation and achieving broader scalar support.

moven0831 commented:
Hi @yaroslavyaroslav @FoodChain1028, good news. I’ve successfully reproduced MSM on the CPU using a logic similar to that on the GPU. The code now supports:

- Random EC points from the BN254 base field  
- Random scalars from the BN254 scalar field  
- Random input sizes from 1 to 65,535 (this sets `chunk_size = 4`, haven’t tried the case where `chunk_size = 16`).  
  *(Note: I only recommend testing input sizes from 1 to 10 to verify correctness on the GPU pipeline.)*

Main corrections include:
1. Fixed SMVP logic in final point accumulation  
2. Adjusted PBPR logic to align with signed index logic (Reference: [Zprize 2023 writeup](https://github.com/z-prize/2023-entries/tree/main/prize-2-msm-wasm/webgpu-only/tal-derei-koh-wei-jie#which-steps-need-to-be-modified))  
3. Corrected bit shift calculation for the last 256-bit chunk  
4. Fixed indexing in Horner’s Method

Next step:
1. Help finalize the current GPU pipeline
2. Improve the CPU version with parallel execution

FoodChain1028 reviewed (approved):
Overall LGTM, just one question: what was the issue before, thank you : )

moven0831 commented:
> Overall LGTM, just one question: what was the issue before, thank you : )

To be more specific,

1. Fixed SMVP logic in final point accumulation 
https://github.com/zkmopro/gpu-acceleration/blob/8a0ec1a5734ae4f85e0cad3c4624042435137d9f/mopro-msm/src/msm/metal_msm/cuzk_cpu_reproduction.rs#L499-L511

2. Adjusted PBPR logic to align with signed index logic (Reference: [Zprize 2023 writeup](https://github.com/z-prize/2023-entries/tree/main/prize-2-msm-wasm/webgpu-only/tal-derei-koh-wei-jie#which-steps-need-to-be-modified))
https://github.com/zkmopro/gpu-acceleration/blob/8a0ec1a5734ae4f85e0cad3c4624042435137d9f/mopro-msm/src/msm/metal_msm/cuzk_cpu_reproduction.rs#L560-L561

3. Corrected bit shift calculation for the last 256-bit chunk
https://github.com/zkmopro/gpu-acceleration/blob/8a0ec1a5734ae4f85e0cad3c4624042435137d9f/mopro-msm/src/msm/metal_msm/cuzk_cpu_reproduction.rs#L311-L313

4. Fixed indexing in Horner’s Method
https://github.com/zkmopro/gpu-acceleration/blob/8a0ec1a5734ae4f85e0cad3c4624042435137d9f/mopro-msm/src/msm/metal_msm/cuzk_cpu_reproduction.rs#L176-L177
